### PR TITLE
default to libc++ for Clang

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
This PR includes three commits (semantically separated, please rebase-merge):

1. default to libc++ when using Clang
   * fixed stdlibc++ vs. libc++ inconsistencies (notably `std::vector<bool>` which showed up in the qa_grc and plugin system handling)
   * explicit setting type to pmtv::pmt conversion (Clang/libc++ seems to promote bool to in)
   * boys-scout-rule: fixed some low-hanging-fruit compiler warning

2. added `lifecycle::StateMachine`
   * performs a hard/signaling check for block and scheduler state-transitions
   * see in-code qa_Lifecycle and LifeCycle doc for details

3. lifecycle::StateMachine integration into Block<T>
   * replaced unprotected state changes in `Block<T>` with the `lifecycle::StateMachine`
   * fixed various state transition errors in `Block<T>` and other blocks that implemented custom state transitions. 
   * added emitting of message in case the transitions are violated 
   * intercept exceptions of user-implemented lifecycle methods and signal this via the `msgOut` message port
   
*N.B. for reviewing it's recommended to review commit-by-commit.*

### CODE REVIEW

File: blocks/basic/include/gnuradio-4.0/basic/common_blocks.hpp [hunk 1]
```diff
@@ -59,7 +59,7 @@ ENABLE_REFLECTION_FOR_TEMPLATE(builtin_counter, in, out);
 //  - use Block::set_name instead of returning an empty name
 // TODO: Inherit from Block class when create new block.
 template<typename T>
-class multi_adder : public gr::BlockModel {
+class multi_adder : public gr::lifecycle::StateMachine<multi_adder<T>>, public gr::BlockModel {
```
Comment (drslebedev) [inline blocks/basic/include/gnuradio-4.0/basic/common_blocks.hpp:62]: In the next iteration I will rewrite this Block to inherit directly from 'Block<T>'

**Review by drslebedev — APPROVED** [review]
Thank you for submitting this PR!
Commit 1: Compiler Compatibility and Minor Fixes
Looks good to me.

Commit 2 & 3: lifecycle::StateMachine Introduction and Integration
The overall design and integration into the existing codebase are very well executed. 
The introduction of lifecycle::StateMachine significantly enhances the handling and robustness of LifeCycle state transitions and checks. 

The documentation is clear, detailed, and helpful. 

The integration with the newly introduced `MessagePort` for error handling looks really good, serving as a prime example for future use cases.
The unit tests are comprehensive and provide good coverage!

In conclusion, I didn't encounter any major issues with the code. It's ready for merging as is. Great work!
